zarr/storage.py

```diff
@@ -1395,13 +1395,19 @@ def __getitem__(self, key):
     def setitems(self, values):
         if self.mode == 'r':
             raise ReadOnlyError()
-        values = {self._normalize_key(key): val for key, val in values.items()}
+
+        # Normalize keys and make sure the values are bytes
+        values = {
+            self._normalize_key(key): ensure_contiguous_ndarray_or_bytes(val)
+            for key, val in values.items()
+        }
```

Comment:
Something we may consider is delaying the instantiation of the `dict` so these copies occur as these values are requested. This can be a follow-on though